src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -2377,11 +2378,12 @@ private unsafe void CreateUriInfo(Flags cF)
         Done:
             cF |= Flags.MinimalUriInfoSet;
             info.DnsSafeHost = _dnsSafeHost;
-            lock (_string)
+
+            Interlocked.CompareExchange(ref _info, info, null!);
+            lock (_info)
```

Comment:
I'd feel more comfortable with this if we didn't have several `_info = null!;` statements strewn about.  While it may not actually be possible, if one of those were to execute, it could cause us to lock on a different instance, as another thread races to initialize the field.  Can we get rid of all of those, instead just initializing `_info = null!` where it's declared (and then subsequently getting rid of that entirely when we get compiler support for https://github.com/dotnet/runtime/issues/31877)?

Reply:
`_info` is being reset to null only in constructors (more specifically methods that are only called by constructors). It is being reset when we try to reuse the `this` instance when merging base and relative Uris - I suppose this was done to potentially save a Uri allocation in those ctors.

I don't think we can easily remove such call sites if we are reusing the same instance for different inputs.

Reply:
> _info is being reset to null only in constructors (more specifically methods that are only called by constructors).

In other such cases, we've done:
```C#
private Whatever _info = null!;
```
when declaring the field.  My question was why is that inappropriate here?

If we end up initializing it to non-null in every constructor, then I'd expect the above should be sufficient.

If we don't, then it would seem the field should be nullable rather than non-nullable.

> It is being reset when we try to reuse the this instance when merging base and relative Uris

Are you saying it's becoming non-null and then null again?  That's exactly what I'm concerned about.

Reply:
> In other such cases, we've done `private Whatever _info = null!;`

This is what is done in Uri as well.

> If we end up initializing it to non-null in every constructor, then I'd expect the above should be sufficient.

After the constructor `_info` _may_ be set. I believe it will only be initialized by the constructor if there is non-ascii in the input.

> If we don't, then it would seem the field should be nullable rather than non-nullable.

Idealy, yes. But it would create a lot of noise throughout Uri without #31877 (could be avoided by using the value returned by EnsureUriInfo).

> Are you saying it's becoming non-null and then null again? That's exactly what I'm concerned about. Is it that by the end of the ctor it'll always end up being non-null, regardless of whether it transitions between null and non-null multiple times in the ctor?

Yes, in the constructor. It could be initialized in EnsureUriInfo, cleared in the constructor, and initialized again to different input.
```c#
private void CreateUri(Uri baseUri, string? relativeUri, bool dontEscape)
{
    // Parse relativeUri and populate Uri internal data.
    CreateThis(relativeUri, dontEscape, UriKind.RelativeOrAbsolute);

    // ... potentially return or throw ...

    // relativeUri has been updated to merged base + relative

    _flags = Flags.Zero;
    _info = null!;
    _syntax = null!;
    // If not resolved, we reparse modified Uri string and populate Uri internal data.
    CreateThis(relativeUri, dontEscape, UriKind.Absolute);
}
```
From what I can tell, it means that:
a) for simple single-input ctors, it is either left null or initialized
b) for base + relative ctors, it is either never initialized **or** it is initialized and then potentially cleared, and initialized again to a different input.

In any case, an instance of Uri when the ctor returns should be safe to use in multiple threads (as far as `_info` is concerned).

Reply:
> Is it that by the end of the ctor it'll always end up being non-null, regardless of whether it transitions between null and non-null multiple times in the ctor?

No - it will be null until EnsureUriInfo is lazily called. That may happen in the constructor or it may happen when other members that require more parsing info are accessed.

Reply:
> or it may happen when other members that require more parsing info are accessed.

In that case, the field should be nullable.

Reply:
> In that case, the field should be nullable.

Is this PR the right place for that? I believe it's worth waiting for #31877.

Reply:
> Is this PR the right place for that?

No.

Can you help me understand why we need this lock at all? e.g. why can't we use a compare-exchange loop to update the flags?

Reply:
I was looking into that but we use the lock in other places when updating flags as well. [One tricky case](https://github.com/dotnet/runtime/blob/129ee5252757ec8bdfccf9b271ded36675570309/src/libraries/System.Private.Uri/src/System/Uri.cs#L2398-L2410) is where we lock around a whole block of custom parser logic. I am also 99% sure we are not consistent about using the lock when updating the flags.

This PR was meant to be a relatively non-invasive change to fix the lock-on-input-string behavior, there is more work to be done to make Uri actually thread-safe.

Reply:
> I was looking into that but we use the lock in other places when updating flags as well

But this is the only place in the code where it was locking on this string instance, right? If so, this was never synchronized at all with those other places.

> This PR was meant to be a relatively non-invasive change

Understood. I'm just pushing on it because changing locking is not something to be taken lightly, and it's not clear from a quick code inspection that it's correct. If other places in the code are already locking on this instance, then it's also potentially introducing more contention.

Reply:
>But this is the only place in the code where it was locking on this string instance, right?

Correct, this method was only synchronized to itself

Reply:
> Correct, this method was only synchronized to itself

So if it's only synchronizing with itself, what is the concern with a compare-exchange loop?

> I am also 99% sure we are not consistent about using the lock when updating the flags.

I suggest we revisit all of the locking altogether, rather than trying to do it piecemeal.  This locking on a string instance has existed forever, and to my knowledge we've never received an actual complaint about it from a developer, just from static analysis tools.

Reply:
@stephentoub While working on lock-less thread safety here, I hit an issue with `Interlocked.Or` as I assumed it returns the original value.
It seems that your implementation in #32216 implements `And` and `Or` to return the **new value** and not the **original value** as was described in the proposal (#23819) and as [`InterlockedAnd`](https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedand#return-value) and [`InterlockedOr`](https://docs.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedor#return-value) windows APIs and [intrinsics](https://docs.microsoft.com/en-gb/previous-versions/b11125ze(v=vs.85)) behave. 

Am I correct that this is an implementation bug and not the desired behavior? If so I can PR a fix before updating this PR.

cc @GrabYourPitchforks

Reply:
It returns the new value, just as do Interlocked.Increment, Interlocked.Decrement, and Interlocked.Add. I think it'd be very confusing if Interlocked.Add and Interlocked.Or were not consistent.

Reply:
Under Behavior in #23819

> Like their arithmetic counterparts, these APIs perform the operation atomically and return the original value to the caller.

> I assume the reason is that you can't always reconstruct the original value from the new?

It can be used to determine if a flag change was made or not.

Every other implementation of atomic/interlocked I can find (windows API, visual c++ intrinsics, Rust Atomic) have the behavior of fetch_and / fetch_or - returning the original value.

Reply:
> It can be used to determine if a flag change was made or not.

Right, that's why I said "reconstruct the original value". You can do so with Add, Increment, and Decrement. You can't necessarily with And and Or.

Reply:
Ok, I'll change it when I get to my desk.

Reply:
Sorry, I just saw that I was mentioned on this. That was a typo in my original issue description - my bad. :(